Add tutorial with manual sparsity pattern (#209)
JuliaSmoothOptimizers/ADNLPModels.jl: docs/src/sparsepattern.md

# Improve sparse derivatives

> In this tutorial, we show a simple trick to potentially improve the computation of sparse Jacobian and Hessian matrices.

Reviewer: ```suggestion
In this tutorial, we show a feature of ADNLPModels.jl to potentially improve the computation of sparse Jacobian and Hessian.
```

\end{aligned}
```

Using a simple quadrature formula for the objective functional and a forward finite difference for the differential equation, one can obtain a finite-dimensional continuous optimization problem.
One implementation is available in the package [`OptimizationProblems.jl`](https://github.com/JuliaSmoothOptimizers/OptimizationProblems.jl).

```@example ex1
using ADNLPModels
using SparseArrays

T = Float64
n = 100000
N = div(n, 2)
h = 1 // N
x0 = 1
gamma = 3
function f(y; N = N, h = h)
  @views x, u = y[1:N], y[(N + 1):end]
  return 1 // 2 * h * sum((u[k] - 1) * x[k] + (u[k + 1] - 1) * x[k + 1] for k = 1:(N - 1))
end
function c!(cx, y; N = N, h = h, gamma = gamma)
  @views x, u = y[1:N], y[(N + 1):end]
  for k = 1:(N - 1)
    cx[k] = x[k + 1] - x[k] - 1 // 2 * h * gamma * (u[k] * x[k] + u[k + 1] * x[k + 1])
  end
  return cx
end
lvar = vcat(-T(Inf) * ones(T, N), zeros(T, N))
uvar = vcat(T(Inf) * ones(T, N), ones(T, N))
xi = vcat(ones(T, N), zeros(T, N))
lcon = ucon = vcat(one(T), zeros(T, N - 1))

@elapsed begin
  nlp = ADNLPModel!(f, xi, lvar, uvar, [1], [1], T[1], c!, lcon, ucon; hessian_backend = ADNLPModels.EmptyADbackend)
end

```

> `ADNLPModel` will automatically try to prepare AD-backend to compute sparse Jacobian and Hessian.

Reviewer: ```suggestion
`ADNLPModel` will automatically prepare an AD backend for computing sparse Jacobian and Hessian.
```

We disabled the Hessian computation here to focus the measurement on the Jacobian computation.

> The keyword argument `show_time = true` can also be passed to the problem's constructor to get more detailed information about the time used to prepare the AD backend.

Reviewer: On devrait s'en servir pour isoler le calcul du sparsity pattern / coloring / preparation AD / AD / decompression.
Peut-être ouvrir une issue?

Author: Bonne idée!

```@example ex1
using NLPModels
x = sqrt(2) * ones(n)
jac_nln(nlp, x)
```

> However, it can be rather costly to determine for a given function the sparsity pattern of the Jacobian and the Lagrangian Hessian matrices.

Reviewer: ```suggestion
However, it can be rather costly to determine for a given function the sparsity pattern of the Jacobian and the Hessian of the Lagrangian.
```

> The good news is that it can be quite easy to have a good approximation of this pattern dealing with problems like our optimal control investment problem, and problem with differential equations in the constraints in general.

Reviewer: ```suggestion
The good news is that determining this pattern a priori can be relatively straightforward, especially for problems like our optimal control investment problem and other problems with differential equations in the constraints.
```

The following example instantiates the Jacobian backend while manually providing the sparsity pattern.

```@example ex2
using ADNLPModels
using SparseArrays

T = Float64
n = 100000
N = div(n, 2)
h = 1 // N
x0 = 1
gamma = 3
function f(y; N = N, h = h)
  @views x, u = y[1:N], y[(N + 1):end]
  return 1 // 2 * h * sum((u[k] - 1) * x[k] + (u[k + 1] - 1) * x[k + 1] for k = 1:(N - 1))
end
function c!(cx, y; N = N, h = h, gamma = gamma)
  @views x, u = y[1:N], y[(N + 1):end]
  for k = 1:(N - 1)
    cx[k] = x[k + 1] - x[k] - 1 // 2 * h * gamma * (u[k] * x[k] + u[k + 1] * x[k + 1])
  end
  return cx
end
lvar = vcat(-T(Inf) * ones(T, N), zeros(T, N))
uvar = vcat(T(Inf) * ones(T, N), ones(T, N))
xi = vcat(ones(T, N), zeros(T, N))
lcon = ucon = vcat(one(T), zeros(T, N - 1))

@elapsed begin

>   J = spzeros(Bool, N - 1, n)

Reviewer: Il faut que tu fasses `colptr`, `rowval`, `nzval` toi même et qu'ensuite tu utilises le constructeur `SparseMatrixCSC`.
Tu fais nnz allocations dynamique avec à chaque fois un check si le nnz existe déjà.
Ça ne peut être que lent :smiley:

  for i =1:(N - 1)
    J[i, i] = true
    J[i, i + 1] = true
    J[i, N + i] = true
    J[i, N + i + 1] = true
  end
  jac_back = ADNLPModels.SparseADJacobian(n, f, N - 1, c!, J)
  nlp = ADNLPModel!(f, xi, lvar, uvar, [1], [1], T[1], c!, lcon, ucon; hessian_backend = ADNLPModels.EmptyADbackend, jacobian_backend = jac_back)
end
```

> A similar Jacobian matrix is obtained potentially at a lower price.

Reviewer: ```suggestion
We recover the same Jacobian.
```

Reviewer: C'est pas vraiment à plus bas prix vu que c'est que de la différentiation automatique pour calculer le jacobien.
C'est vraiment sur le "setup" du backend qu'on gagne.

```@example ex2
using NLPModels
x = sqrt(2) * ones(n)
jac_nln(nlp, x)
```

> The same can be done for the Lagrangian Hessian.

Reviewer: ```suggestion
The same can be done for the Hessian of the Lagrangian.
```
